File: src/main/java/net/lizhao/scriq/eval/Value.java
package net.lizhao.scriq.eval;

import java.math.BigDecimal;
import java.util.concurrent.CompletableFuture;

public class Value {
    public static Value VOID = new Value(new Object());

    final Object value;

    public Value(Object value) {
        this.value = value;
    }

    public Boolean asBoolean() {
        return (Boolean)value;
    }

    public BigDecimal asBigDecimal() {
        return (BigDecimal)value;
    }

    public String asString() {
        return String.valueOf(value);
    }

    public CompletableFuture<Object> asFuture() { return (CompletableFuture<Object>)value; }

    public boolean isBigDecimal() {
        return value instanceof BigDecimal;
    }

    public boolean isString() {
        return value instanceof String;
    }

    public boolean isFuture() { return value instanceof CompletableFuture; }

    @Override
    public int hashCode() {

        if(value == null) {
            return 0;
        }

        return this.value.hashCode();
    }

    @Override
    public boolean equals(Object o) {

        if(value == o) {
            return true;
        }

        if (this.isBigDecimal()) {
            if (o instanceof Integer
                    || o instanceof Long
                    || o instanceof Short
                    || o instanceof Byte) {
                return this.value.equals(BigDecimal.valueOf(((Number) o).longValue()));
            } else if (o instanceof BigDecimal) {
                return this.value.equals(o);
            } else if (o instanceof String) {
                return this.value.equals(new BigDecimal((String) o));
            } else if (o instanceof  Double){
                return this.value.equals(BigDecimal.valueOf(((Number) o).doubleValue()));
            } else {
                return this.value.equals(((Value)o).value);
            }
        }

        if (this.isString()) {
            return this.value.equals(o);
        }

        if(value == null || o == null || o.getClass() != this.getClass()) {
            return false;
        }

        return this.value.equals(((Value)o).value);
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }
}

File: src/main/java/net/lizhao/scriq/eval/Value.kt
package net.lizhao.scriq.eval

import net.lizhao.scriq.py.Python3Parser
import org.jetbrains.kotlinx.multik.ndarray.data.*
import java.util.concurrent.CompletableFuture

class Value(val value: Any?) {
    fun asInt(): Int? {
        return (value as Double).toInt()
    }
    fun asBoolean(): Boolean? {
        return value as Boolean?
    }

    fun asDouble(): Double? {
        return value as Double?
    }

    fun asString(): String {
        return value.toString()
    }

    fun asList(): List<Any> {
        return listOf(value) as List<Any>
    }

    fun asFuture(): CompletableFuture<Any>? {
        return value as CompletableFuture<Any>?
    }

    fun asNDA(): NDA {
        return value as NDA
    }

    val isDouble: Boolean
        get() = value is Double
    val isString: Boolean
        get() = value is String
    val isFuture: Boolean
        get() = value is CompletableFuture<*>
    val isList: Boolean
        get() = value is List<*>
    val isNDA: Boolean
        get() =  value is NDA

    override fun hashCode(): Int {
        return value?.hashCode() ?: 0
    }

    override fun equals(o: Any?): Boolean {
        if (value === o) {
            return true
        }
        if (isDouble) {
            return if (o is Number)
                value == o.toDouble()
            else if (o is String)
                value == o.toDouble()
            else
                value == (o as Value?)!!.value
        }
        if (isString) {
            return value == o
        }
        return if (value == null || o == null || o.javaClass != this.javaClass)
            false
        else
            value == (o as Value).value
    }

    operator fun rem(right: Value): Value {
        return Value(right.asDouble()?.let { value as Double % it })
    }
    operator fun plus(right: Value): Value {
        return when (value) {
            is Double -> if (right.value is Double) Value (value + right.value) else Value((value as Double).toString() + right.value.toString())
            is String -> Value(value as String + right)
            is NDA -> Value(value as NDA + right)
            else -> throw RuntimeException("'+' doesn't support the type'")
        }
    }

    operator fun minus(right: Value): Value {
        return when (value) {
            is Double -> Value(right.asDouble()?.let { value as Double - it })
            is NDA -> Value(value as NDA - right)
            else -> throw RuntimeException("'-' doesn't support the type'")
        }
    }

    operator fun div(right: Value): Value {
        return when (value) {
            is Double -> Value(right.asDouble()?.let { value as Double / it })
            is NDA -> Value(value as NDA / right)
            else -> throw RuntimeException("'/' doesn't support the type'")
        }
    }

    operator fun times(right: Value): Value {
        return when (value) {
            is Double -> Value(right.asDouble()?.let { value as Double * it })
            is NDA -> Value(value as NDA * right)
            else -> throw RuntimeException("'*' doesn't support the type'")
        }
    }

    override fun toString(): String {
        return value.toString()
    }

    companion object {
        var VOID = Value(Any())
    }
}
